### PR TITLE
feat(metrics): Add metrics buffering and sampling

### PR DESCRIPTION
**Pull request by tonyo**
This PR adds the following features: 
1. Buffering + queuing option (`metrics.buffering`). If enabled, metrics will be queued and buffered in a separate thread. Set to `true` by default.
Caveat: `BufferedUdpMetricSink` doesn't flush until the buffer is full, so _theoretically_, if relay doesn't emit any metrics for a while (or if the sampling rate is set to something low), there could be delays in sending the metrics. But some metrics relay sends (e.g. about open/reused connections) are emitted regularly, so it shouldn't be an issue.
2. Global sampling rate (`metrics.sample_rate`). Set to 1.0 by default.

Note: the sampling implementation is pretty naive now, so for example we might simply ignore some calls to counters without scaling it back.

### CODE REVIEW

File: relay-common/src/metrics.rs [hunk 1]
```diff
@@ -98,11 +108,38 @@ impl MetricsClient {
...
 
-        metric.send();
+        match metric.try_send() {
+            Ok(_) => (),
+            Err(error) => log::error!(
+                "Error sending a metric: {}, maximum capacity: {}",
+                LogError(&error),
+                METRICS_MAX_QUEUE_SIZE
+            ),
+        };
```
Comment (flub) [inline relay-common/src/metrics.rs]: `metric.try_send().map_err(|error| ...)` might be more idiomatic

File: relay-common/src/metrics.rs [hunk 2]
```diff
@@ -140,15 +178,51 @@ pub fn configure_statsd<A: ToSocketAddrs>(
...
+        1.0
+    } else if sample_rate <= 0.0 {
+        0.0
+    } else {
+        sample_rate
+    };
+    log::debug!(
+        "metrics sample rate is set to {}{}",
+        normalized_sample_rate,
+        if normalized_sample_rate == 0.0 {
```
Comment (flub) [inline relay-common/src/metrics.rs]: I don't think clippy likes comparing `==` to floats, and rightly so probably.  Probably should use something like `< f32::MIN_POSITIVE` (or `f64`, whichever this is)

File: relay-common/src/metrics.rs [hunk 3]
```diff
@@ -140,15 +178,51 @@ pub fn configure_statsd<A: ToSocketAddrs>(
...
+            ", no metrics will be reported"
+        } else {
+            ""
+        }
+    );
+
+    let socket = UdpSocket::bind("0.0.0.0:0").unwrap();
+    socket.set_nonblocking(true).unwrap();
```
Comment (flub) [inline relay-common/src/metrics.rs:201]: why is this needed?  This is a behaviour change from the current code which can't be undone with simply configuration.

**Review by flub — COMMENTED** [review]
Looks good.  IIRC queued means that cadence starts a new thread and runs the udp sending code there?

File: relay-common/src/metrics.rs (hunk 2 above)
Comment (flub) [reply relay-common/src/metrics.rs]: apparently clippy is happy, so i'm probably just making stuff up

File: relay-common/src/metrics.rs (hunk 1 above)
Comment (jan-auer) [reply relay-common/src/metrics.rs]: Or since we don't require the result any further even:

```rust
if let Err(error) = metric.try_send() {
  log::error!(...);
}
```

File: relay-common/src/metrics.rs (hunk 3 above)
Comment (tonyo) [reply relay-common/src/metrics.rs:201]: It's not the behavior change: https://docs.rs/cadence/0.22.0/src/cadence/client.rs.html#645
We're doing the same, just not inside the `from_udp_host` function

**Review by flub — APPROVED** [review]

File: relay-common/src/metrics.rs [hunk 4]
```diff
@@ -75,6 +83,8 @@ pub struct MetricsClient {
     pub statsd_client: StatsdClient,
     /// Default tags to apply to every metric
     pub default_tags: BTreeMap<String, String>,
+    /// Global sample rate
+    pub sample_rate: f32,
```
Comment (jan-auer) [inline relay-common/src/metrics.rs:87]: Note: After merging, this needs to be added to docs as well.

File: Makefile [hunk 5]
```diff
@@ -16,7 +16,7 @@ clean:
 # Builds
 
 build: setup-git
-	@cargo +stable build --all-features
+	cargo +stable build --all-features
```
Comment (jan-auer) [inline Makefile:19]: This change is unrelated, can we revert?

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,7 @@
 **Features**:
 
 - Rename upstream retries histogram metric and add upstream requests duration metric. ([#816](https://github.com/getsentry/relay/pull/816))
+- Add metrics buffering and sampling. ([#821](https://github.com/getsentry/relay/pull/821))
```
Comment (jan-auer) [inline CHANGELOG.md]: Can you please add a mention of the config option used to tweak this?

File: relay-common/src/metrics.rs [hunk 7]
```diff
@@ -140,15 +177,51 @@ pub fn configure_statsd<A: ToSocketAddrs>(
     prefix: &str,
     host: A,
     default_tags: BTreeMap<String, String>,
+    buffering: bool,
+    sample_rate: f32,
```
Comment (jan-auer) [inline relay-common/src/metrics.rs:181]: This function is now sufficiently complex to become a builder. I'm happy to make the changes in a follow-up, since I'd like to move all metrics and logging related concerns from relay-common into a separate crate anyway.

File: relay-common/src/metrics.rs [hunk 8]
```diff
@@ -140,15 +177,51 @@ pub fn configure_statsd<A: ToSocketAddrs>(
...
     }
-    let statsd_client = StatsdClient::from_udp_host(prefix, &addrs[..]).unwrap();
+
+    let normalized_sample_rate = if sample_rate >= 1.0 {
+        1.0
+    } else if sample_rate <= 0.0 {
+        0.0
+    } else {
+        sample_rate
+    };
```
Comment (jan-auer) [inline relay-common/src/metrics.rs]: _nit_: You can rewrite this as 

```rust
let sample_rate = sample_rate.max(0.).min(1.);
```

Note how I'm shadowing the input parameter to avoid it accidentally being used in place of the normalized value.

File: relay-common/src/metrics.rs (hunk 2 above)
Comment (jan-auer) [reply relay-common/src/metrics.rs]: Comparing to `0.0` is exact, and since we don't do arithmetic on the floats, this should be fine.

**Review by jan-auer — APPROVED** [review]
For a more generic solution it would be great to _eventually_ flush the queued metrics automatically. I agree that this is not a blocker for Relay at the moment.

File: Makefile (hunk 5 above)
Comment (tonyo) [reply Makefile:19]: I found it useful to see what we're actually building, it's the same behavior we have for `test` and `lint` targets.

**Review by jan-auer — APPROVED** [review]
Thanks, @tonyo! Let's follow up with a PR to sentry-docs introducing the new configuration options.
